packages/eslint-plugin/src/rules/no-duplicate-type-constituents.ts

```diff
@@ -145,63 +139,71 @@ export default createRule<Options, MessageIds>({
-      const bracketAfterTokens = context.sourceCode.getTokensAfter(
-        duplicateConstituent.duplicated,
-        { count: bracketBeforeTokens.length },
+        node,
       );
-      const reportLocation: TSESTree.SourceLocation = {
-        start: duplicateConstituent.duplicated.loc.start,
-        end:
-          bracketAfterTokens.length > 0
-            ? bracketAfterTokens[bracketAfterTokens.length - 1].loc.end
-            : duplicateConstituent.duplicated.loc.end,
-      };
+      const bracketAfterTokens = sourceCode.getTokensAfter(node, {
+        count: bracketBeforeTokens.length,
+      });
       context.report({
-        data: {
-          type:
-            parentNode.type === AST_NODE_TYPES.TSIntersectionType
-              ? 'Intersection'
-              : 'Union',
-          previous: context.sourceCode.getText(
-            duplicateConstituent.duplicatePrevious,
-          ),
+        data,
+        messageId,
+        node,
+        loc: {
+          start: node.loc.start,
+          end: (bracketAfterTokens.at(-1) ?? node).loc.end,
         },
-        messageId: 'duplicate',
-        node: duplicateConstituent.duplicated,
-        loc: reportLocation,
-        fix: fixer => {
-          return [
+        fix: fixer =>
+          [
             beforeUnionOrIntersectionToken,
             ...bracketBeforeTokens,
-            duplicateConstituent.duplicated,
+            node,
             ...bracketAfterTokens,
-          ].map(token => fixer.remove(token));
-        },
+          ].map(token => fixer.remove(token)),
       });
     }
     return {
-      ...(!ignoreIntersections && {
-        TSIntersectionType: checkDuplicate,
-      }),
+      ...(!ignoreIntersections && { TSIntersectionType: checkDuplicate }),
       ...(!ignoreUnions && {
-        TSUnionType: checkDuplicate,
+        TSUnionType(node): void {
+          checkDuplicate(node);
+          const maybeTypeAnnotation = node.parent;
+          if (maybeTypeAnnotation.type === AST_NODE_TYPES.TSTypeAnnotation) {
+            const maybeIdentifier = maybeTypeAnnotation.parent;
+            if (
+              maybeIdentifier.type === AST_NODE_TYPES.Identifier &&
+              maybeIdentifier.optional
+            ) {
+              const maybeFunction = maybeIdentifier.parent;
+              const { type } = maybeFunction;
+              if (
+                (type === AST_NODE_TYPES.ArrowFunctionExpression ||
+                  type === AST_NODE_TYPES.FunctionDeclaration ||
+                  type === AST_NODE_TYPES.FunctionExpression) &&
+                maybeFunction.params.includes(maybeIdentifier)
+              ) {
+                const explicitUndefined = node.types.find(
+                  ({ type }) => type === AST_NODE_TYPES.TSUndefinedKeyword,
+                );
+                if (explicitUndefined) {
+                  report('unnecessary', explicitUndefined);
+                }
+              }
+            }
+          }
+        },
```

Comment:
[Bug] This case isn't reported

```ts
function a(arg?: number | (string | undefined)) {}
```

[playground](https://deploy-preview-9479--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.ts&code=GYVwdgxgLglg9mABAQwBTIE4HMD8AuRMEAWwCMBTDRAH0VQGcoMYwsbFwATc4F8zgJQDEAbwC%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNZjGyAQ3yJaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=&tokens=false)

Reply:
This is not currently handled even in the existing rule functionality — we'd need to implement some recursion handling for this. I've gone ahead and opened #9496 for further discussion